modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/CustomWebIdentityTokenCredentialsProviderTests.java

```diff
@@ -125,4 +131,32 @@ public void testCreateWebIdentityTokenCredentialsProvider() throws Exception {
             httpServer.stop(0);
         }
     }
+
+    public void testSupportRegionalizedEndpoints() throws Exception {
+        Map<String, String> environmentVariables = Map.of(
+            "AWS_WEB_IDENTITY_TOKEN_FILE",
+            "/var/run/secrets/eks.amazonaws.com/serviceaccount/token",
+            "AWS_ROLE_ARN",
+            ROLE_ARN,
+            "AWS_STS_REGIONAL_ENDPOINTS",
+            "regional",
+            "AWS_REGION",
+            "us-west-2"
+        );
+        Map<String, String> systemProperties = Map.of();
+
+        var webIdentityTokenCredentialsProvider = new S3Service.CustomWebIdentityTokenCredentialsProvider(
+            getEnvironment(),
+            environmentVariables::get,
+            systemProperties::getOrDefault,
+            Clock.systemUTC()
+        );
+        // We can't verify that webIdentityTokenCredentialsProvider's STS client uses the "https://sts.us-west-2.amazonaws.com"
+        // endpoint in a unit test. The client depends on hardcoded RegionalEndpointsOptionResolver that in turn depends
+        // on the system environment that we can't change in the test. So we just verify we that we called `withRegion`
+        // on stsClientBuilder which should internally correctly configure the endpoint when the STS client is built.
+        assertEquals("us-west-2", webIdentityTokenCredentialsProvider.getStsRegion());
```

Comment:
Is it possible to get a real client instance and check if its sts endpoint really points where it supposed to?

Reply:
I did check that in a manual test. The problem that I can't do that in a unit test because that requires setting the `AWS_STS_REGIONAL_ENDPOINTS` environment variable. I can set environment variables in the `yamlRestTestSTS` test created from Gradle, but I can't verify that the client would sent requests to the real regional endpoint since `yamlRestTestSTS` works with a mocked S3 service.

Reply:
We're mocking STS too I think, and using `com.amazonaws.sdk.stsMetadataServiceEndpointOverride` to specify the STS service endpoint? Could we run this same test without that system property, instead using the `AWS_STS_REGIONAL_ENDPOINTS` env var to specify the mock endpoint?

Reply:
I don't think that will work because custom endpoints and regional endpoints are mutually exclusive. If set `AWS_STS_REGIONAL_ENDPOINTS` in the `yamlRestTestSTS` test, we can't point the S3Client to a `S3Fixture` with mocked credentials. It will use the real regional endpoint like `https://ap-east-1.amazonaws.com`, so we can't successfully authenticate the client.

Reply:
Ah right I see, we're not mocking name resolution enough to do this yet. Would it be possible to do that?